### PR TITLE
fix:windows file permission

### PR DESCRIPTION
**Pull request by Animesh404**
**What kind of change does this PR introduce?**

a bug fix for windows file permission

**Summary**
image

added a retry mechanism (max_retries and retry_delay_seconds) to handle cases where the file deletion fails due to the file being locked or accessed by another process.
you can find `do_fastsam` function in `openadapt/adapters/ultralytics.py`

**Checklist**
* [x] My code follows the style guidelines of OpenAdapt
* [x] I have performed a self-review of my code
* [ ] If applicable, I have added tests to prove my fix is functional/effective
* [x] I have linted my code locally prior to submission
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation (e.g. README.md, requirements.txt)
* [ ] New and existing unit tests pass locally with my changes

**How can your code be run and tested?**

`python -m openadapt.replay VisualReplayStrategy --instructions "<your instructions>"`

### CODE REVIEW

File: openadapt/adapters/ultralytics.py [hunk 1]
```diff
@@ -4,6 +4,8 @@
 """
 # flake8: noqa: E402
 
+import errno
+import time
```
Comment (abrichr) [inline openadapt/adapters/ultralytics.py:8]: I believe you will get a `black` or `flake8` error, these should be ordered alphabetically with `import os`.

File: openadapt/adapters/ultralytics.py [hunk 2]
```diff
@@ -135,8 +137,29 @@ def do_fastsam(
         logger.info(f"{annotation.path=}")
         segmented_image_path = Path(tmp_dir) / result_name
         segmented_image = Image.open(segmented_image_path)
-        os.remove(segmented_image_path)
-
+        # os.remove(segmented_image_path)
```
Comment (abrichr) [inline openadapt/adapters/ultralytics.py:140]: What do you think about removing this?

File: openadapt/adapters/ultralytics.py [hunk 3]
```diff
@@ -135,8 +137,29 @@ def do_fastsam(
...
+
+        # Ensure the image is fully loaded before deletion
+        segmented_image.load()
```
Comment (abrichr) [inline openadapt/adapters/ultralytics.py:143]: Can you please clarify why this is necessary?

File: openadapt/adapters/ultralytics.py [hunk 4]
```diff
@@ -135,8 +137,29 @@ def do_fastsam(
...
+
+        # Attempt to delete the file with retries and delay
+        max_retries = 5
+        retry_delay = 0.1  # seconds
```
Comment (abrichr) [inline openadapt/adapters/ultralytics.py:147]: How about `retry_delay_seconds = 0.1`?

Comment (abrichr) [reply openadapt/adapters/ultralytics.py:147]: Also what do you think about making these keyword arguments to this function with default values?

**Review by abrichr — CHANGES_REQUESTED** [review]
Thank you @Animesh404!! 🙏 🚀 

I left a few comments. Happy to chat about any of it if you like!

Also please paste stack traces as plain text so that they are searchable 🙏

File: openadapt/adapters/ultralytics.py (hunk 3 above)
Comment (Animesh404) [reply openadapt/adapters/ultralytics.py:143]: Some operating systems and file systems lock files while they are being read or processed. Deleting a file that is still being accessed leads to errors or incomplete operations. Ensuring that the image is fully loaded before deletion prevents such issues.

File: openadapt/adapters/ultralytics.py (hunk 2 above)
Comment (Animesh404) [reply openadapt/adapters/ultralytics.py:140]: I've added this in a while loop where I attempt to delete file with retries and delay

File: openadapt/adapters/ultralytics.py (hunk 4 above)
Comment (Animesh404) [reply openadapt/adapters/ultralytics.py:147]: yes it's better that way, I'll update this.

File: openadapt/adapters/ultralytics.py (hunk 2 above)
Comment (abrichr) [reply openadapt/adapters/ultralytics.py]: Right, so why keep this commented out line?

File: openadapt/adapters/ultralytics.py (hunk 3 above)
Comment (abrichr) [reply openadapt/adapters/ultralytics.py:145]: Thank you. Can you please summarize in your comment?

File: openadapt/adapters/ultralytics.py (hunk 4 above)
Comment (abrichr) [reply openadapt/adapters/ultralytics.py]: Once you commit and push a change addressing a comment, please reply to the comment with the commit hash, or at least indicate with a 👍 that it is completed

File: openadapt/adapters/ultralytics.py (hunk 3 above)
Comment (Animesh404) [reply openadapt/adapters/ultralytics.py:146]: commit `bc2d583ee74feb2550137c9a54cc6b184121a59d`

File: openadapt/adapters/ultralytics.py (hunk 1 above)
Comment (Animesh404) [reply openadapt/adapters/ultralytics.py:8]: commit `436af727ccc966d15ecd33ed13af1a05d2b32e20`

File: openadapt/adapters/ultralytics.py (hunk 2 above)
Comment (Animesh404) [reply openadapt/adapters/ultralytics.py]: removed in `436af727ccc966d15ecd33ed13af1a05d2b32e20`

File: openadapt/adapters/ultralytics.py (hunk 1 above)
Comment (abrichr) [reply openadapt/adapters/ultralytics.py:8]: Typically we keep it alphabetical:

```
...
from tempfile import TemporaryDirectory
import os
...
```

Comment (Animesh404) [reply openadapt/adapters/ultralytics.py]: 2259ed5c21ff5f0ea3e5acf3673b81d615614c3a

**Comment by abrichr** [comment]
Looks like the build is failling (clck "Details"):

https://github.com/OpenAdaptAI/OpenAdapt/actions/runs/9628215324/job/26557412749?pr=791

image


```
Run poetry run black --preview --check . --exclude '/(alembic|\.cache|\.venv|venv|contrib|__pycache__)/'
would reformat /Users/runner/work/OpenAdapt/OpenAdapt/OpenAdapt/openadapt/adapters/ultralytics.py
would reformat /Users/runner/work/OpenAdapt/OpenAdapt/OpenAdapt-2024-06-22_22-43-54/adapters/ultralytics.py
```

**Comment by Animesh404** [comment]
> Looks like the build is failling (clck "Details"):
> 
> https://github.com/OpenAdaptAI/OpenAdapt/actions/runs/9628215324/job/26557412749?pr=791
> 
> image
> ```
> Run poetry run black --preview --check . --exclude '/(alembic|\.cache|\.venv|venv|contrib|__pycache__)/'
> would reformat /Users/runner/work/OpenAdapt/OpenAdapt/OpenAdapt/openadapt/adapters/ultralytics.py
> would reformat /Users/runner/work/OpenAdapt/OpenAdapt/OpenAdapt-2024-06-22_22-43-54/adapters/ultralytics.py
> ```

@abrichr this 5bc685fedcdb7bc44a540d67e90a6f11152e0c40 should fix build issues 

> Looks like the build is failling (clck "Details"):
> 
> https://github.com/OpenAdaptAI/OpenAdapt/actions/runs/9628215324/job/26557412749?pr=791
> 
> image
> ```
> Run poetry run black --preview --check . --exclude '/(alembic|\.cache|\.venv|venv|contrib|__pycache__)/'
> would reformat /Users/runner/work/OpenAdapt/OpenAdapt/OpenAdapt/openadapt/adapters/ultralytics.py
> would reformat /Users/runner/work/OpenAdapt/OpenAdapt/OpenAdapt-2024-06-22_22-43-54/adapters/ultralytics.py
> ```

@abrichr this 5bc685fedcdb7bc44a540d67e90a6f11152e0c40 should fix build issues

**Comment by Animesh404** [comment]
> Looks like the build is failling (clck "Details"):
> 
> https://github.com/OpenAdaptAI/OpenAdapt/actions/runs/9628215324/job/26557412749?pr=791
> 
> image
> ```
> Run poetry run black --preview --check . --exclude '/(alembic|\.cache|\.venv|venv|contrib|__pycache__)/'
> would reformat /Users/runner/work/OpenAdapt/OpenAdapt/OpenAdapt/openadapt/adapters/ultralytics.py
> would reformat /Users/runner/work/OpenAdapt/OpenAdapt/OpenAdapt-2024-06-22_22-43-54/adapters/ultralytics.py
> ```

@abrichr this 5bc685fedcdb7bc44a540d67e90a6f11152e0c40 should fix build issue

File: openadapt/adapters/ultralytics.py [hunk 5]
```diff
@@ -82,22 +74,16 @@ def do_fastsam(
...
 
     # Run inference on image
     everything_results = model(
-        image,
-        device=device,
-        retina_masks=retina_masks,
-        imgsz=imgsz,
-        conf=conf,
-        iou=iou,
+        image, device=device, retina_masks=retina_masks, imgsz=imgsz, conf=conf, iou=iou
```
Comment (abrichr) [inline openadapt/adapters/ultralytics.py:86]: This should not be necessary. Please make sure you are using the recommended black/flake8 configs

File: openadapt/adapters/ultralytics.py [hunk 6]
```diff
@@ -80,6 +83,8 @@ def do_fastsam(
     conf: float = 0.4,
     # discards all overlapping boxes with IoU > iou_threshold
     iou: float = 0.9,
+    max_retries: int = 5,
+    retry_delay_seconds: float = 0.1,
```
Comment (abrichr) [inline openadapt/adapters/ultralytics.py:87]: I thought we had implemented exponential backoff, is that still here?

**Review by abrichr — APPROVED** [review]
LGTM!

Edit: @Animesh404 please fix merge conflicts 🙏

**Pull request merged**